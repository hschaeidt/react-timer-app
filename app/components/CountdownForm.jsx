import { h, Component } from 'preact';

export default class CountdownForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      seconds: '',
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  onSubmit(e) {
    e.preventDefault();
    const strSeconds = this.state.seconds;

    if (strSeconds.match(/^[0-9]+$/)) {
      this.setState({
        seconds: '',
      });
      this.props.onSetCountdown(parseInt(strSeconds, 10));
    }
  }

  onChange(e) {
    this.setState({
      seconds: e.target.value,
    });
  }

  render() {
    const { seconds } = this.state;

    return (
      <form onSubmit={this.onSubmit} className="countdown-form">
        <input
          type="text"
          placeholder="Enter time in seconds"
          onChange={this.onChange}
          value={seconds}
        />
        <button className="button expanded">Start</button>
      </form>
    );
  }
}
