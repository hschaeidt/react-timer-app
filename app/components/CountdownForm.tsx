import * as React from 'react';

interface CountdownFormProps {
  onSetCountdown: (seconds: number) => void,
}

interface CountdownFormState {
  seconds: string,
}

export default class CountdownForm extends React.Component<CountdownFormProps, CountdownFormState> {
  constructor(props: CountdownFormProps) {
    super(props);

    this.state = {
      seconds: '',
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  onSubmit(e: React.FormEvent<HTMLFormElement>): void {
    e.preventDefault();
    const strSeconds = this.state.seconds;

    if (strSeconds.match(/^[0-9]+$/)) {
      this.setState({
        seconds: '',
      });
      this.props.onSetCountdown(parseInt(strSeconds, 10));
    }
  }

  onChange(e: React.FormEvent<HTMLInputElement>): void {
    this.setState({
      seconds: e.currentTarget.value,
    });
  }

  render(): JSX.Element {
    const { seconds } = this.state;

    return (
      <form onSubmit={this.onSubmit} className="countdown-form">
        <input
          type="text"
          placeholder="Enter time in seconds"
          onChange={this.onChange}
          value={seconds}
        />
        <button className="button expanded">Start</button>
      </form>
    );
  }
}
