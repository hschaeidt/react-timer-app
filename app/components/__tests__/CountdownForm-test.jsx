import React from 'react';
import { mount } from 'enzyme';
import CountdownForm from '../CountdownForm';

describe('<CountdownForm />', () => {
  it('should exist', () => {
    expect(CountdownForm).toBeDefined();
  });

  it('should call onSetCountdown if valid seconds entered', () => {
    const spy = jest.fn();
    const countdownForm = mount(<CountdownForm onSetCountdown={spy} />);
    countdownForm.find('input').first().simulate('change', { target: { value: '109' } });
    countdownForm.simulate('submit');
    expect(spy).toHaveBeenCalledWith(109);
  });

  it('should not call onSetCountdown if invalid value entered', () => {
    const spy = jest.fn();
    const countdownForm = mount(<CountdownForm onSetCountdown={spy} />);
    countdownForm.find('input').first().simulate('change', { target: { value: '109b' } });
    countdownForm.simulate('submit');
    expect(spy).not.toHaveBeenCalled();
  });
});
