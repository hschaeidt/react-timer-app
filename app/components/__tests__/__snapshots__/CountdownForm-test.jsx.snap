exports[`<CountdownForm /> renders correctly 1`] = `
<form
  className="countdown-form"
  onSubmit={[Function]}>
  <input
    onChange={[Function]}
    placeholder="Enter time in seconds"
    type="text"
    value="" />
  <button
    className="button expanded">
    Start
  </button>
</form>
`;
